pkg/querier/http.go

```diff
@@ -126,3 +126,50 @@ func (q *Querier) LabelHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 }
+
+// TailHandler is a http.HandlerFunc for handling tail queries.
+func (q *Querier) TailHandler(w http.ResponseWriter, r *http.Request) {
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		level.Error(util.Logger).Log("Error in upgrading websocket", fmt.Sprintf("%v", err))
+		return
+	}
+
+	defer func() {
+		err := conn.Close()
+		level.Error(util.Logger).Log("Error closing websocket", fmt.Sprintf("%v", err))
+	}()
+
+	queryRequest, err := httpRequestToQueryRequest(r)
+	if err != nil {
```

Comment:
Should we be parsing the request before the upgrade, so we can return `http.Error(...)`?

Reply:
I tried that actually but client from logcli was not getting proper error message, it was getting bad handshake error.
I will see if I can something.

Reply:
Got it working. Clients can read error code and response body if connection fails, so using same for writing error and status code.